lab3van/util-linux/sys-utils/fstrim.8
<<<<<<< HEAD
.\" -*- nroff -*-
.TH FSTRIM 8 "November 2010" "util-linux" "System Administration"
=======
.TH FSTRIM 8 "July 2014" "util-linux" "System Administration"
>>>>>>> master-vanilla
.SH NAME
fstrim \- discard unused blocks on a mounted filesystem
.SH SYNOPSIS
.B fstrim
<<<<<<< HEAD
=======
.RB [ \-a ]
>>>>>>> master-vanilla
.RB [ \-o
.IR offset ]
.RB [ \-l
.IR length ]
.RB [ \-m
<<<<<<< HEAD
.IR minimum-free-extent ]
=======
.IR minimum-size ]
>>>>>>> master-vanilla
.RB [ \-v ]
.I mountpoint

.SH DESCRIPTION
.B fstrim
is used on a mounted filesystem to discard (or "trim") blocks which are not in
use by the filesystem.  This is useful for solid-state drives (SSDs) and
thinly-provisioned storage.
.PP
By default,
.B fstrim
will discard all unused blocks in the filesystem.  Options may be used to
modify this behavior based on range or size, as explained below.
.PP
The
.I mountpoint
argument is the pathname of the directory where the filesystem
is mounted.
<<<<<<< HEAD

.SH OPTIONS
The \fIoffset\fR, \fIlength\fR, and \fIminimum-free-extent\fR arguments may be
followed by binary (2^N) suffixes KiB, MiB, GiB, TiB, PiB and EiB (the "iB" is
optional, e.g. "K" has the same meaning as "KiB") or decimal (10^N) suffixes
KB, MB, GB, PB and EB.
.IP "\fB\-h, \-\-help\fP"
Print help and exit.
.IP "\fB\-o, \-\-offset\fP \fIoffset\fP"
Byte offset in filesystem from which to begin searching for free blocks
to discard.  Default value is zero, starting at the beginning of the
filesystem.
.IP "\fB\-l, \-\-length\fP \fIlength\fP"
Number of bytes after starting point to search for free blocks to discard.
If the specified value extends past the end of the filesystem,
.B fstrim
will stop at the filesystem size boundary. Default value extends to the end
of the filesystem.
.IP "\fB\-m, \-\-minimum\fP \fIminimum-free-extent\fP"
=======
.PP
Running
.B fstrim
frequently, or even using
.BR "mount -o discard" ,
might negatively affect the lifetime of poor-quality SSD devices.  For most
desktop and server systems the sufficient trimming frequency is once a week.
Note that not all
devices support a queued trim, so each trim command incurs a performance penalty
on whatever else might be trying to use the disk at the time.

.SH OPTIONS
The \fIoffset\fR, \fIlength\fR, and \fIminimum-size\fR arguments may be
followed by the multiplicative suffixes KiB (=1024),
MiB (=1024*1024), and so on for GiB, TiB, PiB, EiB, ZiB and YiB (the "iB"
is optional, e.g., "K" has the same meaning as "KiB") or the suffixes
KB (=1000), MB (=1000*1000), and so on for GB, TB, PB, EB, ZB and YB.

.IP "\fB\-a, \-\-all\fP"
Trim all mounted filesystems on devices that support the discard operation.
The other supplied options, like \fB\-\-offset\fR, \fB\-\-length\fR and
\fB-\-minimum\fR, are applied to all these devices.
Errors from filesystems that do not support the discard operation are silently
ignored.
.IP "\fB\-o, \-\-offset\fP \fIoffset\fP"
Byte offset in the filesystem from which to begin searching for free blocks
to discard.  The default value is zero, starting at the beginning of the
filesystem.
.IP "\fB\-l, \-\-length\fP \fIlength\fP"
The number of bytes (after the starting point) to search for free blocks
to discard.  If the specified value extends past the end of the filesystem,
.B fstrim
will stop at the filesystem size boundary.  The default value extends to
the end of the filesystem.
.IP "\fB\-m, \-\-minimum\fP \fIminimum-size\fP"
>>>>>>> master-vanilla
Minimum contiguous free range to discard, in bytes. (This value is internally
rounded up to a multiple of the filesystem block size).  Free ranges smaller
than this will be ignored.  By increasing this value, the fstrim operation
will complete more quickly for filesystems with badly fragmented freespace,
although not all blocks will be discarded.  Default value is zero, discard
every free block.
.IP "\fB\-v, \-\-verbose\fP"
<<<<<<< HEAD
Verbose execution. When specified 
.B fstrim
will output the number of bytes passed from the filesystem
down the block stack to the device for potential discard. This number is a
maximum discard amount from the storage device's perspective, because
.I FITRIM
ioctl called repeated will keep sending the same sectors for discard repeatedly.

=======
Verbose execution.  With this option
.B fstrim
will output the number of bytes passed from the filesystem
down the block stack to the device for potential discard.  This number is a
maximum discard amount from the storage device's perspective, because
.I FITRIM
ioctl called repeated will keep sending the same sectors for discard repeatedly.
.sp
>>>>>>> master-vanilla
.B fstrim
will report the same potential discard bytes each time, but only sectors which
had been written to between the discards would actually be discarded by the
storage device.  Further, the kernel block layer reserves the right to adjust
the discard ranges to fit raid stripe geometry, non-trim capable devices in a
LVM setup, etc.  These reductions would not be reflected in fstrim_range.len
(the
.B --length
option).
<<<<<<< HEAD
=======
.TP
.BR \-V , " \-\-version"
Display version information and exit.
.TP
.BR \-h , " \-\-help"
Display help text and exit.

.SH RETURN CODES
.IP 0
success
.IP 1
failure
.IP 32
all failed
.IP 64
some filesystem discards have succeeded, some failed
.PP
The command
.B fstrim --all
returns 0 (all succeeded), 32 (all failed) or 64 (some failed, some succeeded).
>>>>>>> master-vanilla

.SH AUTHOR
.nf
Lukas Czerner <lczerner@redhat.com>
Karel Zak <kzak@redhat.com>
.fi
.SH SEE ALSO
.BR mount (8)
.SH AVAILABILITY
The fstrim command is part of the util-linux package and is available from
ftp://ftp.kernel.org/pub/linux/utils/util-linux/.


debian/changelog
util-linux (2.27.160-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- roma <newangel88@mail.ru>  Thu, 24 Dec 2015 13:07:17 +0400


lab3van/util-linux/sys-utils/fstrim.c
/*
 * fstrim.c -- discard the part (or whole) of mounted filesystem.
 *
 * Copyright (C) 2010 Red Hat, Inc. All rights reserved.
 * Written by Lukas Czerner <lczerner@redhat.com>
 *            Karel Zak <kzak@redhat.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 * This program uses FITRIM ioctl to discard parts or the whole filesystem
 * online (mounted). You can specify range (start and length) to be
 * discarded, or simply discard whole filesystem.
 */

#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <fcntl.h>
#include <limits.h>
#include <getopt.h>

#include <sys/ioctl.h>
#include <sys/stat.h>
#include <linux/fs.h>

#include "nls.h"
#include "strutils.h"
#include "c.h"
<<<<<<< HEAD
=======
#include "closestream.h"
#include "pathnames.h"
#include "sysfs.h"
#include "exitcodes.h"

#include <libmount.h>
>>>>>>> master-vanilla

#ifndef FITRIM
struct fstrim_range {
	uint64_t start;
	uint64_t len;
	uint64_t minlen;
};
#define FITRIM		_IOWR('X', 121, struct fstrim_range)
#endif

<<<<<<< HEAD
static void __attribute__((__noreturn__)) usage(FILE *out)
{
	fputs(_("\nUsage:\n"), out);
	fprintf(out,
	      _(" %s [options] <mount point>\n"), program_invocation_short_name);

	fputs(_("\nOptions:\n"), out);
	fputs(_(" -h, --help          this help\n"
		" -o, --offset <num>  offset in bytes to discard from\n"
		" -l, --length <num>  length of bytes to discard from the offset\n"
		" -m, --minimum <num> minimum extent length to discard\n"
		" -v, --verbose       print number of discarded bytes\n"), out);

	fputs(_("\nFor more information see fstrim(8).\n"), out);

=======
/* returns: 0 = success, 1 = unsupported, < 0 = error */
static int fstrim_filesystem(const char *path, struct fstrim_range *rangetpl,
			    int verbose)
{
	int fd, rc;
	struct stat sb;
	struct fstrim_range range;

	/* kernel modifies the range */
	memcpy(&range, rangetpl, sizeof(range));

	fd = open(path, O_RDONLY);
	if (fd < 0) {
		warn(_("cannot open %s"), path);
		rc = -errno;
		goto done;
	}
	if (fstat(fd, &sb) == -1) {
		warn(_("stat of %s failed"), path);
		rc = -errno;
		goto done;
	}
	if (!S_ISDIR(sb.st_mode)) {
		warnx(_("%s: not a directory"), path);
		rc = -EINVAL;
		goto done;
	}
	errno = 0;
	if (ioctl(fd, FITRIM, &range)) {
		rc = errno == EOPNOTSUPP || errno == ENOTTY ? 1 : -errno;

		if (rc != 1)
			warn(_("%s: FITRIM ioctl failed"), path);
		goto done;
	}

	if (verbose) {
		char *str = size_to_human_string(
				SIZE_SUFFIX_3LETTER | SIZE_SUFFIX_SPACE,
				(uint64_t) range.len);
		/* TRANSLATORS: The standard value here is a very large number. */
		printf(_("%s: %s (%" PRIu64 " bytes) trimmed\n"),
				path, str, (uint64_t) range.len);
		free(str);
	}

	rc = 0;
done:
	if (fd >= 0)
		close(fd);
	return rc;
}

static int has_discard(const char *devname, struct sysfs_cxt *wholedisk)
{
	struct sysfs_cxt cxt, *parent = NULL;
	uint64_t dg = 0;
	dev_t disk = 0, dev;
	int rc;

	dev = sysfs_devname_to_devno(devname, NULL);
	if (!dev)
		return 1;
	/*
	 * This is tricky to read the info from sys/, because the queue
	 * atrributes are provided for whole devices (disk) only. We're trying
	 * to reuse the whole-disk sysfs context to optimize this stuff (as
	 * system usually have just one disk only).
	 */
	if (sysfs_devno_to_wholedisk(dev, NULL, 0, &disk) || !disk)
		return 1;
	if (dev != disk) {
		if (wholedisk->devno != disk) {
			sysfs_deinit(wholedisk);
			if (sysfs_init(wholedisk, disk, NULL))
				return 1;
		}
		parent = wholedisk;
	}

	rc = sysfs_init(&cxt, dev, parent);
	if (!rc)
		rc = sysfs_read_u64(&cxt, "queue/discard_granularity", &dg);

	sysfs_deinit(&cxt);
	return rc == 0 && dg > 0;
}


static int uniq_fs_target_cmp(
		struct libmnt_table *tb __attribute__((__unused__)),
		struct libmnt_fs *a,
		struct libmnt_fs *b)
{
	return !mnt_fs_streq_target(a, mnt_fs_get_target(b));
}

static int uniq_fs_source_cmp(
		struct libmnt_table *tb __attribute__((__unused__)),
		struct libmnt_fs *a,
		struct libmnt_fs *b)
{
	int eq;

	if (mnt_fs_is_pseudofs(a) || mnt_fs_is_netfs(a) ||
	    mnt_fs_is_pseudofs(b) || mnt_fs_is_netfs(b))
		return 1;

	eq = mnt_fs_streq_srcpath(a, mnt_fs_get_srcpath(b));
	if (eq) {
		const char *aroot = mnt_fs_get_root(a),
			   *broot = mnt_fs_get_root(b);
		if (!aroot || !broot)
			eq = 0;
		else if (strcmp(aroot, broot) != 0)
			eq = 0;
	}

	return !eq;
}

/*
 * fstrim --all follows "mount -a" return codes:
 *
 * 0  = all success
 * 32 = all failed
 * 64 = some failed, some success
 */
static int fstrim_all(struct fstrim_range *rangetpl, int verbose)
{
	struct libmnt_fs *fs;
	struct libmnt_iter *itr;
	struct libmnt_table *tab;
	struct sysfs_cxt wholedisk = UL_SYSFSCXT_EMPTY;
	int cnt = 0, cnt_err = 0;

	mnt_init_debug(0);

	itr = mnt_new_iter(MNT_ITER_BACKWARD);
	if (!itr)
		err(MOUNT_EX_FAIL, _("failed to initialize libmount iterator"));

	tab = mnt_new_table_from_file(_PATH_PROC_MOUNTINFO);
	if (!tab)
		err(MOUNT_EX_FAIL, _("failed to parse %s"), _PATH_PROC_MOUNTINFO);

	/* de-duplicate by mountpoints */
	mnt_table_uniq_fs(tab, 0, uniq_fs_target_cmp);

	/* de-duplicate by source and root */
	mnt_table_uniq_fs(tab, 0, uniq_fs_source_cmp);

	while (mnt_table_next_fs(tab, itr, &fs) == 0) {
		const char *src = mnt_fs_get_srcpath(fs),
			   *tgt = mnt_fs_get_target(fs);
		char *path;
		int rc = 1;

		if (!src || !tgt || *src != '/' ||
		    mnt_fs_is_pseudofs(fs) ||
		    mnt_fs_is_netfs(fs))
			continue;

		/* Is it really accessible mountpoint? Not all mountpoints are
		 * accessible (maybe over mounted by another fylesystem) */
		path = mnt_get_mountpoint(tgt);
		if (path && strcmp(path, tgt) == 0)
			rc = 0;
		free(path);
		if (rc)
			continue;	/* overlaying mount */

		if (!has_discard(src, &wholedisk))
			continue;
		cnt++;

		/*
		 * We're able to detect that the device supports discard, but
		 * things also depend on filesystem or device mapping, for
		 * example vfat or LUKS (by default) does not support FSTRIM.
		 *
		 * This is reason why we ignore EOPNOTSUPP and ENOTTY errors
		 * from discard ioctl.
		 */
		if (fstrim_filesystem(tgt, rangetpl, verbose) < 0)
		       cnt_err++;
	}

	sysfs_deinit(&wholedisk);
	mnt_unref_table(tab);
	mnt_free_iter(itr);

	if (cnt && cnt == cnt_err)
		return MOUNT_EX_FAIL;		/* all failed */
	if (cnt && cnt_err)
		return MOUNT_EX_SOMEOK;		/* some ok */

	return EXIT_SUCCESS;
}

static void __attribute__((__noreturn__)) usage(FILE *out)
{
	fputs(USAGE_HEADER, out);
	fprintf(out,
	      _(" %s [options] <mount point>\n"), program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Discard unused blocks on a mounted filesystem.\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -a, --all           trim all mounted filesystems that are supported\n"), out);
	fputs(_(" -o, --offset <num>  the offset in bytes to start discarding from\n"), out);
	fputs(_(" -l, --length <num>  the number of bytes to discard\n"), out);
	fputs(_(" -m, --minimum <num> the minimum extent length to discard\n"), out);
	fputs(_(" -v, --verbose       print number of discarded bytes\n"), out);

	fputs(USAGE_SEPARATOR, out);
	fputs(USAGE_HELP, out);
	fputs(USAGE_VERSION, out);
	fprintf(out, USAGE_MAN_TAIL("fstrim(8)"));
>>>>>>> master-vanilla
	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
<<<<<<< HEAD
	char *path;
	int c, fd, verbose = 0;
	struct fstrim_range range;
	struct stat sb;

	static const struct option longopts[] = {
	    { "help",      0, 0, 'h' },
=======
	char *path = NULL;
	int c, rc, verbose = 0, all = 0;
	struct fstrim_range range;

	static const struct option longopts[] = {
	    { "all",       0, 0, 'a' },
	    { "help",      0, 0, 'h' },
	    { "version",   0, 0, 'V' },
>>>>>>> master-vanilla
	    { "offset",    1, 0, 'o' },
	    { "length",    1, 0, 'l' },
	    { "minimum",   1, 0, 'm' },
	    { "verbose",   0, 0, 'v' },
	    { NULL,        0, 0, 0 }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
<<<<<<< HEAD
=======
	atexit(close_stdout);
>>>>>>> master-vanilla

	memset(&range, 0, sizeof(range));
	range.len = ULLONG_MAX;

<<<<<<< HEAD
	while ((c = getopt_long(argc, argv, "ho:l:m:v", longopts, NULL)) != -1) {
		switch(c) {
		case 'h':
			usage(stdout);
			break;
		case 'l':
			if (strtosize(optarg, (uint64_t *) &range.len))
				errx(EXIT_FAILURE,
				     _("failed to parse length: %s"), optarg);
			break;
		case 'o':
			if (strtosize(optarg, (uint64_t *) &range.start))
				errx(EXIT_FAILURE,
				     _("failed to parse offset: %s"), optarg);
			break;
		case 'm':
			if (strtosize(optarg, (uint64_t *) &range.minlen))
				errx(EXIT_FAILURE,
				     _("failed to parse minimum extent length: %s"),
				     optarg);
=======
	while ((c = getopt_long(argc, argv, "ahVo:l:m:v", longopts, NULL)) != -1) {
		switch(c) {
		case 'a':
			all = 1;
			break;
		case 'h':
			usage(stdout);
			break;
		case 'V':
			printf(UTIL_LINUX_VERSION);
			return EXIT_SUCCESS;
		case 'l':
			range.len = strtosize_or_err(optarg,
					_("failed to parse length"));
			break;
		case 'o':
			range.start = strtosize_or_err(optarg,
					_("failed to parse offset"));
			break;
		case 'm':
			range.minlen = strtosize_or_err(optarg,
					_("failed to parse minimum extent length"));
>>>>>>> master-vanilla
			break;
		case 'v':
			verbose = 1;
			break;
		default:
			usage(stderr);
			break;
		}
	}

<<<<<<< HEAD
	if (optind == argc)
		errx(EXIT_FAILURE, _("no mountpoint specified."));

	path = argv[optind++];
=======
	if (!all) {
		if (optind == argc)
			errx(EXIT_FAILURE, _("no mountpoint specified"));
		path = argv[optind++];
	}
>>>>>>> master-vanilla

	if (optind != argc) {
		warnx(_("unexpected number of arguments"));
		usage(stderr);
	}

<<<<<<< HEAD
	if (stat(path, &sb) == -1)
		err(EXIT_FAILURE, _("%s: stat failed"), path);
	if (!S_ISDIR(sb.st_mode))
		errx(EXIT_FAILURE, _("%s: not a directory"), path);

	fd = open(path, O_RDONLY);
	if (fd < 0)
		err(EXIT_FAILURE, _("%s: open failed"), path);

	if (ioctl(fd, FITRIM, &range))
		err(EXIT_FAILURE, _("%s: FITRIM ioctl failed"), path);

	if (verbose)
		/* TRANSLATORS: The standard value here is a very large number. */
		printf(_("%s: %" PRIu64 " bytes were trimmed\n"),
						path, (uint64_t) range.len);
	close(fd);
	return EXIT_SUCCESS;
=======
	if (all)
		rc = fstrim_all(&range, verbose);
	else {
		rc = fstrim_filesystem(path, &range, verbose);
		if (rc == 1) {
			warnx(_("%s: the discard operation is not supported"), path);
			rc = EXIT_FAILURE;
		}
	}

	return rc;
>>>>>>> master-vanilla
}
